components/authz-service/engine/opa/opa_v2_internal_test.go

```diff
@@ -0,0 +1,988 @@
+		engine.Pair{Resource: "system:license", Action: "system:license:request"},
+		engine.Pair{Resource: "notifications:rules", Action: "notifications:notifyRules:list"},
+		engine.Pair{Resource: "iam:roles", Action: "iam:roles:create"},
+	}
+
+	policies, roles := v2BaselinePoliciesAndRoles()
+
+	s.v2p1Store = inmem.NewFromObject(map[string]interface{}{
+		"policies": policies,
+		"roles":    roles,
+	})
+
+	teamCount := []int{0, 1, 10, 30, 50, 100, 150, 300, 500, 1000, 10000}
+	for _, count := range teamCount {
+		b.Run(fmt.Sprintf("with %d teams in input", count), func(b *testing.B) {
+			var resp []engine.Pair
+			var err error
+			for n := 0; n < b.N; n++ {
+				resp, err = s.V2FilterAuthorizedPairs(ctx, append([]string{"user:local:test@example.com"}, randomTeams(count)...), pairs)
+				if err != nil {
+					b.Error(err)
+				}
+			}
+			filteredPairsResp = resp
+		})
+	}
+}
+
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_0_teams_in_input-8       18	   71469413 ns/op	   27715819 B/op	  416006 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_1_teams_in_input-8       15	   70341373 ns/op	   28104209 B/op	  423505 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_10_teams_in_input-8      14	   79268379 ns/op	   31604062 B/op	  490993 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_30_teams_in_input-8      13	  115535168 ns/op	   39381388 B/op	  640964 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_50_teams_in_input-8       9	  126016309 ns/op	   47169624 B/op	  790930 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_100_teams_in_input-8      7	  178359130 ns/op	   66610725 B/op	 1166059 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_150_teams_in_input-8      4	  290961881 ns/op	   86099896 B/op	 1552118 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_300_teams_in_input-8      3	  410043023 ns/op	  144511557 B/op	 2710281 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_500_teams_in_input-8      2	  623028756 ns/op	  222406888 B/op	 4254489 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_1000_teams_in_input-8     1	 1253675227 ns/op	  417184776 B/op	 8115021 allocs/op
+// BenchmarkFilterAuthorizedPairsRealWorldExample/with_10000_teams_in_input-8    1	19126847294 ns/op	 3931882536 B/op	77604823 allocs/op
+// 12/16/19 summary: up to 1-2 seconds with 1000 teams
+
+// Q: Which type of input is computed faster, generic Go interface or specific OPA Term?
+func BenchmarkV2GenericInput(b *testing.B) {
+	var r ast.Value
+	var err error
+	for n := 0; n < b.N; n++ {
+		// always record the result to prevent the compiler eliminating the function
+		// call.
+		r, err = genericV2Input(
+			[]string{"user:local:alice@example.com", "team:local:admins"},
+			"some:resource",
+			"some:resource:action",
+			[]string{"project-1", "project-2"})
```

Comment:
Out of pure stupid curiosity, I wonder if there is value in being pedantic and lifting the construction of these arrays out of of the inner loop here like the are in the V2SpecificInput. Might also be interesting to move the code in BenchmarkV2SpecificInput into a function similar to genericV2Input just so we make sure we are know more precisely what we are benchmarking.

One hopes that the go compiler would make all this pointless since the code should optimize somewhat similarly, but might just also be good so that we can visually confirm "oh, this right here is what the benchmark tests."

Reply:
💡 good point, I hadn't thought about that